File: src/main/java/de/hdawg/rankinginfo/api/v1/PlayersApiController.java
package de.hdawg.rankinginfo.api.v1;

import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.security.SecurityRequirement;
import org.springframework.http.HttpStatus;
import org.springframework.http.ProblemDetail;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import de.hdawg.rankinginfo.api.v1.dto.PlayerDetailData;
import de.hdawg.rankinginfo.api.v1.dto.PlayerDetailResponse;
import de.hdawg.rankinginfo.api.v1.dto.PlayerLink;
import de.hdawg.rankinginfo.api.v1.dto.PlayerSearchItem;
import de.hdawg.rankinginfo.api.v1.dto.PlayerSearchRequest;
import de.hdawg.rankinginfo.api.v1.dto.PlayerSearchResponse;
import de.hdawg.rankinginfo.api.v1.dto.RankingEntry;
import de.hdawg.rankinginfo.domain.RankingCoding;
import de.hdawg.rankinginfo.service.PlayerService;

@RestController
@RequestMapping("/api/v1/players")
@SecurityRequirement(name = "bearerAuth")
public class PlayersApiController {

  private final PlayerService playerService;

  public PlayersApiController(PlayerService playerService) {
    this.playerService = playerService;
  }

  @Operation(summary = "Search players by lastname and optional year of birth")
  @ApiResponse(
      responseCode = "200",
      content = @Content(schema = @Schema(implementation = PlayerSearchResponse.class)))
  @ApiResponse(
      responseCode = "400",
      content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
  @ApiResponse(
      responseCode = "404",
      content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
  @GetMapping
  public ResponseEntity<Object> search(
      @RequestParam(required = false, defaultValue = "") String lastname,
      @RequestParam(required = false, defaultValue = "") String yob) {

    if (lastname.isBlank()) {
      return ResponseEntity.badRequest()
          .body(
              ProblemDetail.forStatusAndDetail(
                  HttpStatus.BAD_REQUEST, "lastname parameter required"));
    }

    if (!yob.isBlank() && yob.length() != 4) {
      return ResponseEntity.badRequest()
          .body(ProblemDetail.forStatusAndDetail(HttpStatus.BAD_REQUEST, "yob must be a 4-digit year"));
    }

    int yy = yob.isBlank() ? 0 : Integer.parseInt(yob.substring(2, 4));
    var players =
        (!yob.isBlank())
            ? playerService.findPlayersByLastnameAndYob(
                lastname,
                yy + RankingCoding.GENDER_FACTOR_JUNIOREN,
                yy + RankingCoding.GENDER_FACTOR_JUNIORINNEN)
            : playerService.findPlayersByLastname(lastname);

    if (players.isEmpty()) {
      return ResponseEntity.status(404)
          .body(ProblemDetail.forStatusAndDetail(HttpStatus.NOT_FOUND, "Not Found"));
    }

    var data =
        players.stream()
            .map(
                p ->
                    new PlayerSearchItem(
                        p.dtbId(),
                        p.lastname(),
                        p.firstname(),
                        p.club(),
                        new PlayerLink("/api/v1/players/" + p.dtbId())))
            .toList();

    return ResponseEntity.ok(
        new PlayerSearchResponse(
            new PlayerSearchRequest(lastname, yob, players.size()), data));
  }

  @Operation(summary = "Get player profile with full ranking history")
  @ApiResponse(
      responseCode = "200",
      content = @Content(schema = @Schema(implementation = PlayerDetailResponse.class)))
  @ApiResponse(
      responseCode = "404",
      content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
  @GetMapping("/{id}")
  public ResponseEntity<Object> show(@PathVariable int id) {
    var rankings = playerService.findNonAggregateRankings(id);
    if (rankings.isEmpty()) {
      return ResponseEntity.status(404)
          .body(ProblemDetail.forStatusAndDetail(HttpStatus.NOT_FOUND, "Not Found"));
    }
    var current = rankings.getFirst();
    var rankingEntries =
        rankings.stream()
            .map(
                r ->
                    new RankingEntry(
                        r.date(), r.ageGroup(), r.rankingPosition(), r.score()))
            .toList();

    return ResponseEntity.ok(
        new PlayerDetailResponse(
            new PlayerDetailData(
                current.dtbId(),
                current.lastname(),
                current.firstname(),
                current.nationality(),
                current.club(),
                current.federation(),
                rankingEntries)));
  }
}

File: src/main/kotlin/de/hdawg/rankinginfo/api/v1/PlayersApiController.kt
package de.hdawg.rankinginfo.api.v1

import de.hdawg.rankinginfo.repository.RankingRepository
import de.hdawg.rankinginfo.service.PlayerService
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/**
 * REST controller exposing player search and player detail endpoints under `/api/v1/players`.
 *
 * All endpoints require a valid Bearer token (see [de.hdawg.rankinginfo.api.security.BearerTokenFilter]).
 */
@RestController
@RequestMapping("/api/v1/players")
@SecurityRequirement(name = "bearerAuth")
class PlayersApiController(
    private val playerService: PlayerService,
    private val rankingRepository: RankingRepository,
) {
    /**
     * Searches players by last name, optionally filtered by year of birth.
     *
     * @param lastname player's last name (required); returns 400 if blank
     * @param yob four-digit year of birth (e.g. "2005"); optional
     * @return [PlayerSearchResponse] on success, [ErrorResponse] on 400/404
     */
    @Operation(summary = "Search players by lastname and optional year of birth")
    @ApiResponses(
        ApiResponse(responseCode = "200", content = [Content(schema = Schema(implementation = PlayerSearchResponse::class))]),
        ApiResponse(responseCode = "400", content = [Content(schema = Schema(implementation = ErrorResponse::class))]),
        ApiResponse(responseCode = "404", content = [Content(schema = Schema(implementation = ErrorResponse::class))]),
    )
    @GetMapping
    fun search(
        @RequestParam(required = false) lastname: String?,
        @RequestParam(required = false) yob: String?,
    ): ResponseEntity<Any> {
        if (lastname.isNullOrBlank()) {
            return ResponseEntity.badRequest().body(ErrorResponse("lastname parameter required"))
        }
        val players =
            if (!yob.isNullOrBlank()) {
                val yobMale = yob.substring(2, 4).toInt() + 100
                playerService.findPlayersByLastnameAndYob(lastname, yobMale, yobMale + 100)
            } else {
                playerService.findPlayersByLastname(lastname)
            }
        if (players.isEmpty()) {
            return ResponseEntity.status(404).body(ErrorResponse("Not Found"))
        }
        return ResponseEntity.ok(
            PlayerSearchResponse(
                request = PlayerSearchRequest(lastname = lastname, yob = yob, total_count = players.size),
                data =
                    players.map { p ->
                        PlayerSearchItem(
                            dtb_id = p.dtbId,
                            lastname = p.lastname,
                            firstname = p.firstname,
                            club = p.club,
                            links = PlayerLink(self = "/api/v1/players/${p.dtbId}"),
                        )
                    },
            ),
        )
    }

    /**
     * Returns the full profile and ranking history for a single player.
     *
     * @param id the player's DTB ID
     * @return [PlayerDetailResponse] on success, [ErrorResponse] with 404 if not found
     */
    @Operation(summary = "Get player profile with full ranking history")
    @ApiResponses(
        ApiResponse(responseCode = "200", content = [Content(schema = Schema(implementation = PlayerDetailResponse::class))]),
        ApiResponse(responseCode = "404", content = [Content(schema = Schema(implementation = ErrorResponse::class))]),
    )
    @GetMapping("/{id}")
    fun show(
        @PathVariable id: Int,
    ): ResponseEntity<Any> {
        val rankings =
            rankingRepository
                .findByDtbIdAndYobRankingFalseAndAgeGroupRankingFalseAndYearEndRankingFalseOrderByDateDescAgeGroupAsc(id)
        if (rankings.isEmpty()) {
            return ResponseEntity.status(404).body(ErrorResponse("Not Found"))
        }
        val current = rankings.first()
        return ResponseEntity.ok(
            PlayerDetailResponse(
                data =
                    PlayerDetailData(
                        dtb_id = current.dtbId,
                        lastname = current.lastname,
                        firstname = current.firstname,
                        nationality = current.nationality,
                        club = current.club,
                        federation = current.federation,
                        rankings =
                            rankings.map { r ->
                                RankingEntry(
                                    quarter = r.date,
                                    age_group = r.ageGroup,
                                    ranking_position = r.rankingPosition,
                                    score = r.score,
                                )
                            },
                    ),
            ),
        )
    }
}
